File: backend/src/services/AuditLogService.ts
import pino from 'pino';
import { AuditEvent, AuditLog } from '../types/entities.js';

const logger = pino();

export class AuditLogService {
  constructor(private auditLogRepository: any) {}

  async logEvent(event: AuditEvent): Promise<AuditLog> {
    logger.info(
      `Logging audit event. Type: ${event.eventType}, Category: ${event.eventCategory}, Action: ${event.action}, Status: ${event.status}`
    );

    const startTime = Date.now();

    try {
      const auditLog: Partial<AuditLog> = {
        eventType: event.eventType,
        eventCategory: event.eventCategory,
        logLevel: event.logLevel,
        sourceSystem: event.sourceSystem,
        tenantId: event.tenantId,
        userId: event.userId,
        userEmail: event.userEmail,
        action: event.action,
        resourceType: event.resourceType,
        resourceId: event.resourceId,
        oldValue: event.oldValue,
        newValue: event.newValue,
        status: event.status,
        httpMethod: event.httpMethod,
        httpStatusCode: event.httpStatusCode,
        endpointPath: event.endpointPath,
        ipAddress: event.ipAddress,
        userAgent: event.userAgent,
        errorMessage: event.errorMessage,
        stacktrace: event.stacktrace,
        additionalContext: event.additionalContext,
        durationMs: event.durationMs,
        createdAt: new Date(),
        createdAtEpoch: BigInt(Date.now())
      };

      const saved = await this.auditLogRepository.save(auditLog);

      const duration = Date.now() - startTime;
      logger.debug(`Audit event logged successfully. LogID: ${saved.logId}, Duration: ${duration}ms`);

      return saved;
    } catch (e) {
      const duration = Date.now() - startTime;
      logger.error(
        `Failed to log audit event. Type: ${event.eventType}, Duration: ${duration}ms, Error: ${(e as Error).message}`
      );
      throw e;
    }
  }

  async logUserLogin(
    userId: string,
    userEmail: string,
    tenantId: string,
    success: boolean,
    ipAddress: string,
    userAgent: string
  ): Promise<void> {
    logger.info(`Logging user login. UserID: ${userId}, Email: ${userEmail}, TenantID: ${tenantId}, Success: ${success}`);

    const event: AuditEvent = {
      eventType: 'USER_LOGIN',
      eventCategory: 'AUTHENTICATION',
      action: 'LOGIN',
      resourceType: 'USER',
      resourceId: userId,
      userId,
      userEmail,
      tenantId,
      ipAddress,
      userAgent,
      status: success ? 'SUCCESS' : 'FAILURE',
      logLevel: success ? 'INFO' : 'WARN'
    };

    await this.logEvent(event);
  }

  async logUserLogout(userId: string, tenantId: string): Promise<void> {
    logger.info(`Logging user logout. UserID: ${userId}, TenantID: ${tenantId}`);

    const event: AuditEvent = {
      eventType: 'USER_LOGOUT',
      eventCategory: 'AUTHENTICATION',
      action: 'LOGOUT',
      resourceType: 'USER',
      resourceId: userId,
      userId,
      tenantId,
      status: 'SUCCESS',
      logLevel: 'INFO'
    };

    await this.logEvent(event);
  }

  async logConfigChange(resourceId: string, oldValue: string, newValue: string, userId: string): Promise<void> {
    logger.info(`Logging config change. ConfigID: ${resourceId}, UserID: ${userId}`);

    const event: AuditEvent = {
      eventType: 'CONFIG_CHANGED',
      eventCategory: 'CONFIGURATION',
      action: 'UPDATE',
      resourceType: 'CONFIG',
      resourceId,
      oldValue,
      newValue,
      userId,
      status: 'SUCCESS',
      logLevel: 'WARN'
    };

    await this.logEvent(event);
  }

  async logApiCall(
    method: string,
    path: string,
    statusCode: number,
    durationMs: number,
    userId: string,
    ipAddress: string
  ): Promise<void> {
    logger.debug(`Logging API call. Method: ${method}, Path: ${path}, Status: ${statusCode}, Duration: ${durationMs}ms`);

    const logLevel = statusCode >= 400 ? 'WARN' : 'DEBUG';
    const status = statusCode >= 400 ? 'FAILURE' : 'SUCCESS';

    const event: AuditEvent = {
      eventType: 'API_CALL',
      eventCategory: 'API',
      action: method,
      httpMethod: method,
      endpointPath: path,
      httpStatusCode: statusCode,
      userId,
      ipAddress,
      status,
      logLevel,
      durationMs
    };

    await this.logEvent(event);
  }

  async logPermissionDenied(
    userId: string,
    resource: string,
    action: string,
    tenantId: string
  ): Promise<void> {
    logger.warn(`Logging permission denied. UserID: ${userId}, Resource: ${resource}, Action: ${action}, TenantID: ${tenantId}`);

    const event: AuditEvent = {
      eventType: 'PERMISSION_DENIED',
      eventCategory: 'SECURITY',
      action,
      resourceType: resource,
      userId,
      tenantId,
      status: 'FAILURE',
      logLevel: 'WARN',
      errorMessage: `User does not have permission to ${action} ${resource}`
    };

    await this.logEvent(event);
  }

  async logError(errorType: string, errorMessage: string, stacktrace: string, userId: string, resourceId: string): Promise<void> {
    logger.error(`Logging error event. Type: ${errorType}, Message: ${errorMessage}`);

    const event: AuditEvent = {
      eventType: errorType,
      eventCategory: 'ERROR',
      action: 'ERROR',
      resourceId,
      userId,
      status: 'FAILURE',
      logLevel: 'ERROR',
      errorMessage,
      stacktrace
    };

    await this.logEvent(event);
  }

  async getAuditLogsByUser(userId: string): Promise<AuditLog[]> {
    logger.info(`Querying audit logs for user: ${userId}`);
    return this.auditLogRepository.findByUserId(userId);
  }

  async getAuditLogsByTenant(tenantId: string): Promise<AuditLog[]> {
    logger.info(`Querying audit logs for tenant: ${tenantId}`);
    return this.auditLogRepository.findByTenantId(tenantId);
  }

  async getAuditLogsByEventType(eventType: string): Promise<AuditLog[]> {
    logger.info(`Querying audit logs by event type: ${eventType}`);
    return this.auditLogRepository.findByEventType(eventType);
  }

  async getAuditLogsByLevel(logLevel: string): Promise<AuditLog[]> {
    logger.info(`Querying audit logs by level: ${logLevel}`);
    return this.auditLogRepository.findByLogLevel(logLevel);
  }

  async countAuditLogs(): Promise<number> {
    const count = await this.auditLogRepository.count();
    logger.info(`Total audit logs in system: ${count}`);
    return count;
  }
}

File: backend/src/main/java/com/keycloak/saas/audit/AuditLogService.java
package com.keycloak.saas.audit;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.util.List;

/**
 * Service for audit logging.
 * All events are logged to audit_logs table (immutable, never deleted).
 * Used for compliance, security, debugging.
 *
 * Audit logs are critical: enable detailed event tracking for all state-changing operations.
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class AuditLogService {

    private final AuditLogRepository auditLogRepository;

    // ========================================================================
    // LOG OPERATIONS
    // ========================================================================

    /**
     * Log an audit event to database.
     * This is the main entry point for all audit logging.
     *
     * @param event AuditEvent with all event details
     * @return Saved audit log entity
     */
    @Transactional
    public AuditLogEntity logEvent(AuditEvent event) {
        log.info("Logging audit event. Type: {}, Category: {}, Action: {}, Status: {}",
                event.getEventType(), event.getEventCategory(), event.getAction(), event.getStatus());

        long startTime = System.currentTimeMillis();

        try {
            AuditLogEntity auditLog = AuditLogEntity.builder()
                    .eventType(event.getEventType())
                    .eventCategory(event.getEventCategory())
                    .logLevel(event.getLogLevel())
                    .sourceSystem(event.getSourceSystem())
                    .tenantId(event.getTenantId())
                    .userId(event.getUserId())
                    .userEmail(event.getUserEmail())
                    .action(event.getAction())
                    .resourceType(event.getResourceType())
                    .resourceId(event.getResourceId())
                    .oldValue(event.getOldValue())
                    .newValue(event.getNewValue())
                    .status(event.getStatus())
                    .httpMethod(event.getHttpMethod())
                    .httpStatusCode(event.getHttpStatusCode())
                    .endpointPath(event.getEndpointPath())
                    .ipAddress(event.getIpAddress())
                    .userAgent(event.getUserAgent())
                    .errorMessage(event.getErrorMessage())
                    .stacktrace(event.getStacktrace())
                    .additionalContext(event.getAdditionalContext())
                    .createdAt(OffsetDateTime.now(ZoneOffset.UTC))
                    .createdAtEpoch(System.currentTimeMillis())
                    .build();

            AuditLogEntity saved = auditLogRepository.save(auditLog);

            long duration = System.currentTimeMillis() - startTime;
            log.debug("Audit event logged successfully. LogID: {}, Duration: {}ms",
                    saved.getLogId(), duration);

            return saved;
        } catch (Exception e) {
            long duration = System.currentTimeMillis() - startTime;
            log.error("Failed to log audit event. Type: {}, Duration: {}ms, Error: {}",
                    event.getEventType(), duration, e.getMessage(), e);
            throw e;
        }
    }

    /**
     * Log user login event.
     * @param userId User ID
     * @param userEmail User email
     * @param tenantId Tenant ID
     * @param success Whether login succeeded
     * @param ipAddress Client IP address
     * @param userAgent Client user agent
     */
    @Transactional
    public void logUserLogin(String userId, String userEmail, String tenantId,
                            boolean success, String ipAddress, String userAgent) {
        log.info("Logging user login. UserID: {}, Email: {}, TenantID: {}, Success: {}",
                userId, userEmail, tenantId, success);

        AuditEvent event = AuditEventBuilder.builder()
                .eventType("USER_LOGIN")
                .eventCategory("AUTHENTICATION")
                .action("LOGIN")
                .resourceType("USER")
                .resourceId(userId)
                .userId(userId)
                .userEmail(userEmail)
                .tenantId(tenantId)
                .ipAddress(ipAddress)
                .userAgent(userAgent)
                .status(success ? "SUCCESS" : "FAILURE")
                .logLevel(success ? "INFO" : "WARN")
                .build();

        logEvent(event);
    }

    /**
     * Log user logout event.
     * @param userId User ID
     * @param tenantId Tenant ID
     */
    @Transactional
    public void logUserLogout(String userId, String tenantId) {
        log.info("Logging user logout. UserID: {}, TenantID: {}", userId, tenantId);

        AuditEvent event = AuditEventBuilder.builder()
                .eventType("USER_LOGOUT")
                .eventCategory("AUTHENTICATION")
                .action("LOGOUT")
                .resourceType("USER")
                .resourceId(userId)
                .userId(userId)
                .tenantId(tenantId)
                .status("SUCCESS")
                .logLevel("INFO")
                .build();

        logEvent(event);
    }

    /**
     * Log configuration change event.
     * @param resourceId Config ID
     * @param oldValue Previous value
     * @param newValue New value
     * @param userId User making change
     */
    @Transactional
    public void logConfigChange(String resourceId, String oldValue, String newValue, String userId) {
        log.info("Logging config change. ConfigID: {}, UserID: {}", resourceId, userId);

        AuditEvent event = AuditEventBuilder.builder()
                .eventType("CONFIG_CHANGED")
                .eventCategory("CONFIGURATION")
                .action("UPDATE")
                .resourceType("CONFIG")
                .resourceId(resourceId)
                .oldValue(oldValue)
                .newValue(newValue)
                .userId(userId)
                .status("SUCCESS")
                .logLevel("WARN")
                .build();

        logEvent(event);
    }

    /**
     * Log API call (request/response).
     * @param method HTTP method (GET, POST, etc)
     * @param path API endpoint path
     * @param statusCode HTTP response status code
     * @param durationMs Request duration in milliseconds
     * @param userId User making request (if authenticated)
     * @param ipAddress Client IP address
     */
    @Transactional
    public void logApiCall(String method, String path, int statusCode,
                          long durationMs, String userId, String ipAddress) {
        log.debug("Logging API call. Method: {}, Path: {}, Status: {}, Duration: {}ms",
                method, path, statusCode, durationMs);

        String logLevel = statusCode >= 400 ? "WARN" : "DEBUG";
        String status = statusCode >= 400 ? "FAILURE" : "SUCCESS";

        AuditEvent event = AuditEventBuilder.builder()
                .eventType("API_CALL")
                .eventCategory("API")
                .action(method)
                .httpMethod(method)
                .endpointPath(path)
                .httpStatusCode(statusCode)
                .userId(userId)
                .ipAddress(ipAddress)
                .status(status)
                .logLevel(logLevel)
                .durationMs((int) durationMs)
                .build();

        logEvent(event);
    }

    /**
     * Log permission denied event.
     * @param userId User who was denied
     * @param resource Resource being accessed
     * @param action Action attempted
     * @param tenantId Tenant ID
     */
    @Transactional
    public void logPermissionDenied(String userId, String resource, String action, String tenantId) {
        log.warn("Logging permission denied. UserID: {}, Resource: {}, Action: {}, TenantID: {}",
                userId, resource, action, tenantId);

        AuditEvent event = AuditEventBuilder.builder()
                .eventType("PERMISSION_DENIED")
                .eventCategory("SECURITY")
                .action(action)
                .resourceType(resource)
                .userId(userId)
                .tenantId(tenantId)
                .status("FAILURE")
                .logLevel("WARN")
                .errorMessage("User does not have permission to " + action + " " + resource)
                .build();

        logEvent(event);
    }

    /**
     * Log error/exception event.
     * @param errorType Type of error (VALIDATION_ERROR, DATABASE_ERROR, etc)
     * @param errorMessage Human-readable error message
     * @param stacktrace Full stack trace
     * @param userId User associated with error (if any)
     * @param resourceId Resource involved in error (if any)
     */
    @Transactional
    public void logError(String errorType, String errorMessage, String stacktrace,
                        String userId, String resourceId) {
        log.error("Logging error event. Type: {}, Message: {}", errorType, errorMessage);

        AuditEvent event = AuditEventBuilder.builder()
                .eventType(errorType)
                .eventCategory("ERROR")
                .action("ERROR")
                .resourceId(resourceId)
                .userId(userId)
                .status("FAILURE")
                .logLevel("ERROR")
                .errorMessage(errorMessage)
                .stacktrace(stacktrace)
                .build();

        logEvent(event);
    }

    // ========================================================================
    // QUERY OPERATIONS
    // ========================================================================

    /**
     * Get audit logs for a specific user.
     * @param userId User ID
     * @return List of audit log entries
     */
    public List<AuditLogEntity> getAuditLogsByUser(String userId) {
        log.info("Querying audit logs for user: {}", userId);
        return auditLogRepository.findByUserIdOrderByCreatedAtDesc(userId);
    }

    /**
     * Get audit logs for a specific tenant.
     * @param tenantId Tenant ID
     * @return List of audit log entries
     */
    public List<AuditLogEntity> getAuditLogsByTenant(String tenantId) {
        log.info("Querying audit logs for tenant: {}", tenantId);
        return auditLogRepository.findByTenantIdOrderByCreatedAtDesc(tenantId);
    }

    /**
     * Get audit logs by event type.
     * @param eventType Event type to filter by
     * @return List of audit log entries
     */
    public List<AuditLogEntity> getAuditLogsByEventType(String eventType) {
        log.info("Querying audit logs by event type: {}", eventType);
        return auditLogRepository.findByEventTypeOrderByCreatedAtDesc(eventType);
    }

    /**
     * Get audit logs by log level.
     * @param logLevel Log level to filter by (DEBUG, INFO, WARN, ERROR)
     * @return List of audit log entries
     */
    public List<AuditLogEntity> getAuditLogsByLevel(String logLevel) {
        log.info("Querying audit logs by level: {}", logLevel);
        return auditLogRepository.findByLogLevelOrderByCreatedAtDesc(logLevel);
    }

    /**
     * Count total audit logs.
     * @return Total count
     */
    public long countAuditLogs() {
        long count = auditLogRepository.count();
        log.info("Total audit logs in system: {}", count);
        return count;
    }
}
